feat: add translations to public form

siddarth2824 opened the pull request:
## Changes
- Add language control to allow public form users to toggle languages on the form
- Use admin input translations on public form
- Add translations for prevent submission from form logic
- Add fixed translations for some components on the public form which are not generated by admin. For example, labels in `FormInstructionsContainer`. This [sheet](https://docs.google.com/spreadsheets/d/1u_dH53sglH5w7jV83rU5NFIhjk80_FIJ/edit?usp=sharing&ouid=101878657257239680436&rtpof=true&sd=true) contains the fixed translations. Highlighted in green are the translations added in this PR.

## Before & After Screenshots
| Description | Before | After | 
| ----------- | ------- | ----- |
| Language control component to allow users to select the language to view the public form in | | <img width="1512" alt="Screenshot 2024-12-11 at 5 24 24 PM" src="https://github.com/user-attachments/assets/b12235d7-3384-4452-bb84-d74042e72462"> |

siddarth2824 committed:
feat: add model fields to represent translations for form fields (#7457)

* feat: add fields in model to represent translations for form fields

* feat: add shared types to represent form field translations

* fix: use unicode locales

siddarth2824 committed:
feat: add settings tab and toggle for multi lang feature (#7561)

* feat: add settings tab and toggle for multi lang feature

* feat: add fields to joi validations

* fix: refactor code to address comments

* fix: failing tests

* feat: add tooltip over icons

* feat: add chromatic tests

siddarth2824 committed:
feat: add translations on public form

siddarth2824 committed:
fix: missing provider values for preview and template providers

siddarth2824 committed:
feat: add fixed translations for yes and no field

siddarth2824 committed:
chore: fix code

siddarth2824 committed:
chore: remove unused code and add import

siddarth2824 committed:
feat: add fixed translations for est time taken string

siddarth2824 committed:
chore: remove unnecessary loggin

siddarth2824 committed:
feat: add translations for verifiable fields

siddarth2824 committed:
chore: fix tests

siddarth2824 committed:
feat: add error if user adds less than required translations for form fields with options

siddarth2824 committed:
feat: add beta flag for multi language translation feature for admins

siddarth2824 committed:
chore: refactor code

siddarth2824 committed:
feat: add fixed translations for default placeholders on public form

siddarth2824 committed:
feat: add fixed translations for not found label

siddarth2824 committed:
feat: add translations for maximum file size label

siddarth2824 committed:
feat: add translations for others label

siddarth2824 committed:
feat: add fixed translations for add row label

siddarth2824 committed:
feat: add translation for prevent submission messages

siddarth2824 committed:
fix: make prevent submission translations optional

siddarth2824 committed:
fix: tests

LoneRifle commented:
Generally, we would use i18next for text. Would be happy to point out how wthat can be done.

LoneRifle reviewed (dismissed):
Yea I'm afraid I might have to ask you to refactor to incorporate the use of i18next, as we discussed the other time

LoneRifle committed:
refactor: use i18next to toggle languages

Use i18next to hold currently selected language rather than use our own
internal state flag in PublicFormContext. That way, a form submitter
can control the language used across the entire Form service.

- Rework LanguageControl and PublicFormContext to use i18next to hold
  currently selected language, falling back on `Language.ENGLISH`
- Rework components to lookup current lang from `useTranslation()`, not
  `PublicFormContext.selectedPublicFormLanguage` or prop passing

LoneRifle committed:
fix: stub i18next for YesNo,Email,Verify fields

LoneRifle committed:
refactor(i18n): move translations from `fixedTranslations`

LoneRifle committed:
refactor(i18n): extract hard-coded translations

siddarth2824 committed:
feat: add title translations for MyInfo fields

LoneRifle commented on shared/constants/field/myinfo/index.ts:
```diff
@@ -50,6 +53,11 @@ export const types: MyInfoFieldBlock[] = [
       'The registered name of the form-filler. This field is verified by ICA for Singaporeans/PRs & foreigners on Long-Term Visit Pass, and by MOM for Employment Pass holders.',
     fieldType: BasicField.ShortText,
     previewValue: 'PHUA CHU KANG',
+    titleTranslations: [
+      { language: Language.CHINESE, translation: '姓名' },
+      { language: Language.MALAY, translation: 'Nama' },
+      { language: Language.TAMIL, translation: 'பெயர்' },
+    ],
```
Er we've talked about this, can you use i18next please?

siddarth2824 committed:
feat: always default language back to English on CreatePage

KenLSM commented on frontend/src/features/admin-form/create/builder-and-design/utils/fieldCreation.ts:
```diff
@@ -220,11 +220,19 @@ export const getMyInfoFieldCreationMeta = (
     disabled: false,
     required: true,
     title: MYINFO_ATTRIBUTE_MAP[myInfoAttribute].value,
+    titleTranslations: MYINFO_ATTRIBUTE_MAP[myInfoAttribute]?.titleTranslations,
```
Note to self: this is expected to remains as a temporary solution as it is complex to extract the translations to be specified in locales file.

KenLSM commented on frontend/src/features/admin-form/settings/SettingsPage.tsx:
```diff
@@ -51,6 +56,18 @@ export const SettingsPage = (): JSX.Element => {
+          icon: LanguageTranslation,
+          component: SettingsMultiLangPage,
+          path: 'language',
+        }
```
/nit it is easier to read the inverse.

```suggestion
    !isUserLoading &&
    !isFormSettingLoading &&
    user?.betaFlags?.multiLangTranslation
      ? {
          label: 'Multi-language',
          icon: LanguageTranslation,
          component: SettingsMultiLangPage,
          path: 'language',
        }
      : null
```

KenLSM commented on shared/constants/field/myinfo/index.ts:
Note to self: sister file relation to translations within a constants file as interim instead of having in locales.

KenLSM commented on frontend/src/features/public-form/components/FormFields/PublicFormSubmitButton.tsx:
```diff
@@ -65,6 +65,19 @@ export const PublicFormSubmitButton = ({
   }, [formInputs, formFields, formLogics])
 
+  const selectedLanguage = i18n.language as Language
+  const preventSubmissionMessage = useMemo(() => {
```
Possible to use `getValueInSelectedLanguage()`?

KenLSM commented on frontend/src/templates/Field/Radio/RadioField.tsx:
```diff
@@ -50,6 +53,8 @@ export const RadioField = ({
     [disableRequiredValidation, schema],
   )
 
+  const defaultEnglishRadioOptions = schema.fieldOptions
```
/supernit would suggest to not name it as `defaultEnglishRadioOptions` but rather straight up `englishRadioOptions`. 

I had initially through that this value is to be overwritten as `defaultX` suggests that the value would be replaced. However, this is not the case as we strictly only want to pass `englishRadioOptions` value to the backend.

KenLSM commented on frontend/src/templates/Field/Dropdown/DropdownField.tsx:
```diff
@@ -23,12 +25,53 @@ export const DropdownField = ({
+  const selectedLanguage = i18n.language as Language
+
+  const fieldOptions: ComboboxItem[] = useMemo(() => {
+    const defaultEnglishFieldOptions = schema.fieldOptions
```
/supernit in similar style with `RadioField`. `defaultX` suggests that this value would be replaced, but we shouldn't since we only want to pass the english values to the backend.

KenLSM commented on frontend/src/templates/Field/Dropdown/DropdownField.tsx:
```diff
@@ -23,12 +25,53 @@ export const DropdownField = ({
+    // render the default field options in English.
+    if (
+      translationIdx !== -1 &&
+      fieldOptionsTranslations[translationIdx].translation.length ===
```
Check why this change is much more than radio field

KenLSM replied:
Is the reason why we cannot use `getFieldOptionsInSelectedLanguage` is due to the fact that we have a custom extractor[1] here? 

Context of question: I do like the utils function as it abstracts out the selection language out of the component rendering logic, so I'm sadden to see that it cannot be utilised. When one is reading through this component, the complex logic of language selection / detection (such as detection and selection of language + its defaults) is a distraction.

[1] the assumed offending extractor
```
return translatedFieldOptions.map((translatedFieldOption, index) => {
  return {
    value: defaultEnglishFieldOptions[index],
    label: translatedFieldOption,
  }
})
```

If this is the only problem, we can likely look into a utils that supports custom extractors as an enhancement in the future.

KenLSM commented on frontend/src/templates/Field/Checkbox/CheckboxField.tsx:
```diff
@@ -59,11 +62,20 @@ export const CheckboxField = ({
     [disableRequiredValidation, schema],
   )
 
+  const defaultEnglishCheckboxOptions = schema.fieldOptions
```
/supernit in similar style with `RadioField`. `defaultX` suggests that this value would be replaced, but we shouldn't since we only want to pass the english values to the backend.

KenLSM commented on frontend/src/utils/multiLanguage.ts:
```diff
@@ -16,3 +20,64 @@ export const convertUnicodeLocaleToLanguage = (language: Language) => {
+      defaultValue.length
+  ) {
+    return fieldOptionsTranslations[translationIdx].translation
+  } else {
```
/nit this `else` can be omitted. I generally prefer an `if without else` aka guard clauses

KenLSM commented on frontend/src/utils/multiLanguage.ts:
```diff
@@ -16,3 +20,64 @@ export const convertUnicodeLocaleToLanguage = (language: Language) => {
+  translations,
+  selectedLanguage,
+}: SelectedLanguageProps) => {
+  let title = defaultValue
```
/supernit Why `title`? This function is trying to get the `value` but this suggest that the value is used for `titles`

KenLSM commented on frontend/src/templates/Field/Section/SectionField.tsx:
```diff
@@ -39,6 +41,34 @@ export const BaseSectionField = forwardRef<
+  const selectedLanguage = i18n.language as Language
+
+  let title = schema.title
+  const titleTranslations = schema?.titleTranslations ?? []
```
Can use `getValueInSelectedLanguage`?

KenLSM commented on frontend/src/templates/Field/Section/SectionField.tsx:
```diff
@@ -39,6 +41,34 @@ export const BaseSectionField = forwardRef<
+  }
+
+  let description = schema.description
+  const descriptionTranslations = schema?.descriptionTranslations ?? []
```
Can use `getValueInSelectedLanguage`?

KenLSM commented on frontend/src/templates/Field/Table/ColumnCell.tsx:
```diff
@@ -82,12 +84,52 @@ const DropdownColumnCell = ({
+        defaultEnglishFieldOptions.length
+    ) {
+      const translatedFieldOptions =
+        fieldOptionsTranslations[translationIdx].translation
```
Same question as in `frontend/src/templates/Field/Dropdown/DropdownField.tsx` regarding the utilisation of multi-lang utils. Was it due to a need for custom extractor?

KenLSM commented on frontend/src/templates/Field/Table/TableFieldContainer.tsx:
```diff
@@ -21,8 +24,42 @@ export const TableFieldContainer = ({
+    title = titleTranslations[titleTranslationIdx].translation
+  }
+
+  let description = schema.description
```
Can use `getValueInSelectedLanguage`?

KenLSM commented on frontend/src/features/public-form/components/LanguageControl/LanguageControl.tsx:
```diff
@@ -0,0 +1,113 @@
+      px={{ base: '1.5rem', md: 0 }}
+      justifyContent={{ base: 'start', md: 'center' }}
+    >
+      <HStack mt="-32px" bg="white" borderRadius="4px" shadow="md">
```
Able to use `rem`?

KenLSM commented on frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx:
```diff
@@ -345,9 +375,26 @@ export const TranslationListSection = ({
+          {formLogicPreventSubmissions && (
+            <>
+              <QuestionRow
+                key="Form Logic"
```
Is `key` needed here since we're not rendering them in a list like manner.

siddarth2824 replied:
Even though we are not rendering in a list manner, I think its required because there are other `<QuestionRow>` components being rendered at the same nested level as itself. Hence requiring a key to uniquely identify the `<QuestionRow>` components from each other. I keep seeing the error on my console regarding this, hence I added it. I am not very sure if its required, but when I add it, the error on console seems to disappear.

KenLSM commented on frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx:
```diff
@@ -345,9 +375,26 @@ export const TranslationListSection = ({
           {/* End Page Translation */}
           {form.endPage && (
             <QuestionRow
+              key="End Page"
```
Is `key` needed here since we're not rendering them in a list like manner.

KenLSM commented on frontend/src/components/FormEndPage/EndPageBlock.tsx:
```diff
@@ -43,6 +46,31 @@ export const EndPageBlock = ({
+    selectedLanguage,
+  })
+
+  const paragraph = useMemo(() => {
```
Can use `getValueInSelectedLanguage`?

KenLSM reviewed (commented):
Several question. All non-blocking.

siddarth2824 committed:
fix: update myInfo translations and address comments

siddarth2824 committed:
feat: add custom header to track user selected form language

KenLSM reviewed (approved):
LGTM! Thanks for taking time to split the PRs for easier review 👍
